### ISSUE TITLE
Wishlist: Please add laps filter to histogram view

### ISSUE THREAD

**Issue by idealist1508** [alex-hhh/ActivityLog2#64]
I am using ActivityLog2 to compare my trainings.
To compare laps with each other please add laps filter to histogram view:
![image](https://user-images.githubusercontent.com/4608274/110634666-86261c00-81aa-11eb-899f-f24bcd48fb63.png)
What do you think? I could try to implement this and make PR.

**Comment by alex-hhh** [comment]
Yes, I would like to have this feature.  Looking at the screenshot, it looks like you already implemented this?  If so, can you open a pull request for this, so I have a look at the code?

**Comment by idealist1508** [comment]
No, it is only "photoshop". I would try to implement it.

**Comment by alex-hhh** [comment]
I pushed some initial work towards implementing this to the `ah/histogram-laps` branch, https://github.com/alex-hhh/ActivityLog2/tree/ah/histogram-laps.  Note that there were some corresponding changes to the `data-frame` package, so you will need to run a `git submodule update` after you checkout that branch.

This implementation is incomplete.  At least the following are missing:

* once you select a lap, there is no way to select the histogram for the entire activity
* currently the plot X axis only shows the minimum range for the current lap, making laps difficult to compare, the Y axis (height) is also at its minimum range.  It would be good to get the axis range from the histogram of the entire activity , so lap histograms are easier to compare
* even better it would be good to plot the entire histogram in the background faded out.
* as an implementation, this is somewhat inefficient, as it recalculates everything each time a new lap is selected -- this should be changed to only update what is needed, using "plot-state" and "plot-data" structures, as it is done in "inspect-graphs.rkt".

**Comment by idealist1508** [comment]
>     * currently the plot X axis only shows the minimum range for the current lap
I would prefer to allow manual setting of ranges for both axes. This will make it easy to compare even different activities. I am  missing this feature in a chart view too.

**Comment by idealist1508** [comment]
How to debug? I set breakpoints in inspect-histogram.rkt  in a function refresh-plot, but it dosn't hit. What i am doing wrong?
Do you now any tutorials?

**Comment by alex-hhh** [comment]
I don't use the Racket debugger myself, but you can ask for help about it on the racket-users list at https://groups.google.com/g/racket-users.

I use several techniques for debugging, and did not write all of them down, but here are some tips of what might help in this case:

* first, I usually set-up a simpler program, with just the modules I am working on, so it loads faster.  I pushed some changes to the branch, and you can see a sample program for this in the [etc/scratch-histogram-laps.rkt[(https://github.com/alex-hhh/ActivityLog2/blob/ah/histogram-laps/etc/scratch-inspect-histogram.rkt) file.  You will need to replace the session ID with one from your database.   This program will load faster, so you can run it repeatedly more efficiently.
* next, in this case, I would just use printf statements , to print out values that I am interested in.   This is especially useful, since ActivityLog2 uses separate threads, and I am not sure that the Racket debugger can handle them.

Hope this helps.

**Comment by idealist1508** [comment]
I add a code to deselect an item with a click on it. https://github.com/idealist1508/ActivityLog2/tree/ah/histogram-laps 
Please review.

**Comment by alex-hhh** [comment]
Thanks for the suggestion for unclicking items -- I ended up using a slightly different approach which does not make use of a `timer%` object and also unselects the selected highlight from the map and graphs view.  I merged these changes to the master branch, you can check them out.

**Comment by idealist1508** [comment]
This solution has one bug. Double click doesn't work on selected items.

**Comment by idealist1508** [comment]
I found that under linux in a single mode it was possible to deselect an item by a ctrl click.  I don't know if this works on Windows too.

**Comment by alex-hhh** [comment]
Double click works fine on Windows with these changes, and I am not sure how it would stop working in Linux, since the "double click" is a different message....

**Comment by idealist1508** [comment]
Maybe my description was not optimal, so I made a video.
![dblclick](https://user-images.githubusercontent.com/4608274/113576351-b6cd6a00-961f-11eb-8c03-19df0bced934.gif)
My second comment was, that I found how to deselect an Item [when list is in a single mode](https://github.com/alex-hhh/ActivityLog2/blob/36a4bb8af45db19cea02e982e22379acb18d0c49/rkt/widgets/qresults-list.rkt#L383). 
![singlemode](https://user-images.githubusercontent.com/4608274/113577390-7d95f980-9621-11eb-8617-c51adcaa5b02.gif)
If it works on other platforms then a workaround with the multiselect isn't needed.

**Comment by alex-hhh** [comment]
Thanks for explaining this -- un-select is not supported on Windows, but I just reverted this change as it is probably best not to modify the native widget behavior as it is not portable *racket/gui uses the native widget on each platform whenever possible).
